iroh-bytes/src/downloader.rs

```diff
@@ -1084,88 +1053,142 @@ impl<G: Getter<Connection = D::Connection>, D: Dialer> Service<G, D> {
-    fn get_next_provider_hash_for_node(&mut self, node: &NodeId) -> Option<Hash> {
-        let hash = self
-            .provider_hashes_by_node
-            .get(node)
-            .and_then(|hashes| hashes.front())
-            .copied();
-        if let Some(hash) = hash {
-            self.move_hash_to_back(node, hash);
+    /// Register nodes for a hash, but only if the hash is already in our queue.
+    fn add_nodes_if_hash_exists(&mut self, hash: Hash, nodes: impl Iterator<Item = NodeId>) {
+        if let Some(hash_entry) = self.hash_node.get_mut(&hash) {
+            for node in nodes {
+                hash_entry.insert(node);
+                let node_entry = self.node_hash.entry(node).or_default();
+                node_entry.insert(hash);
+            }
         }
-        hash
     }
 
     /// Signal the registry that this hash is no longer of interest.
-    fn remove(&mut self, hash: Hash) {
-        if let Some(nodes) = self.candidates.remove(&hash) {
-            for node in nodes.keys() {
-                if let Some(hashes) = self.provider_hashes_by_node.get_mut(node) {
-                    hashes.retain(|h| *h != hash);
+    fn remove_hash(&mut self, hash: &Hash) {
+        if let Some(nodes) = self.hash_node.remove(hash) {
+            for node in nodes {
+                if let Some(hashes) = self.node_hash.get_mut(&node) {
+                    hashes.remove(hash);
+                    if hashes.is_empty() {
+                        self.node_hash.remove(&node);
+                    }
                 }
             }
         }
     }
 
-    /// Move a hash to the back of the provider queue for a node.
-    fn move_hash_to_back(&mut self, node: &NodeId, hash: Hash) {
-        let hashes = self.provider_hashes_by_node.get_mut(node);
-        if let Some(hashes) = hashes {
-            debug_assert_eq!(hashes.front(), Some(&hash));
-            if !hashes.is_empty() {
-                hashes.rotate_left(1);
+    fn remove_node(&mut self, node: &NodeId) {
+        if let Some(hashes) = self.node_hash.remove(node) {
+            for hash in hashes {
+                if let Some(nodes) = self.hash_node.get_mut(&hash) {
+                    nodes.remove(node);
+                    if nodes.is_empty() {
+                        self.hash_node.remove(&hash);
+                    }
+                }
+            }
+        }
+    }
+
+    fn remove_hash_from_node(&mut self, hash: &Hash, node: &NodeId) {
+        if let Some(nodes) = self.hash_node.get_mut(hash) {
+            nodes.remove(node);
+            if nodes.is_empty() {
+                self.remove_hash(hash);
+            }
+        }
+        if let Some(hashes) = self.node_hash.get_mut(node) {
+            hashes.remove(hash);
+            if hashes.is_empty() {
+                self.remove_node(node);
+            }
+        }
+    }
+}
+
+/// Helper struct to merge multiple tags for a blob.
+#[derive(Debug, Default)]
+struct TagSet {
```

Comment:
Not a suggestion, but a question. Would it make sense to have this around the tag api/db code? seems useful enough and perhaps more pertinent to have there

Reply:
SGTM, moved to `iroh-bytes/src/util.rs`